File: src/lib/utils/diff.ts
// Unified diff → the lines a viewer draws. Parsed here rather than in Rust
// because hunk structure is what the *viewer* needs and nothing else does;
// sending it over IPC would be a second representation of git's own output,
// kept in step by hand.

export type DiffLineKind = "add" | "remove" | "context" | "meta";

export interface DiffLine {
  kind: DiffLineKind;
  text: string;
  /** Line number in the old file, or null for an added line. */
  oldLine: number | null;
  /** Line number in the new file, or null for a removed line. */
  newLine: number | null;
}

export interface DiffStat {
  added: number;
  removed: number;
}

export interface ParsedDiff {
  lines: DiffLine[];
  stat: DiffStat;
  /** True when git produced nothing — an unchanged file, or a binary one. */
  empty: boolean;
  /** Git declined to show content, e.g. `Binary files … differ`. */
  binary: boolean;
}

const HUNK = /^@@ -(\d+)(?:,\d+)? \+(\d+)(?:,\d+)? @@/;

/**
 * Everything before the first `@@` is git's file header — `diff --git`, index
 * hashes, `---`/`+++`. It says nothing the viewer doesn't already show in its own
 * header, so it is dropped rather than rendered as noise above every diff.
 */
export function parseDiff(raw: string): ParsedDiff {
  const lines: DiffLine[] = [];
  const stat: DiffStat = { added: 0, removed: 0 };
  let binary = false;
  let oldLine = 0;
  let newLine = 0;
  let inHunk = false;

  for (const text of raw.split("\n")) {
    const hunk = HUNK.exec(text);
    if (hunk) {
      inHunk = true;
      oldLine = Number(hunk[1]);
      newLine = Number(hunk[2]);
      lines.push({ kind: "meta", text, oldLine: null, newLine: null });
      continue;
    }
    if (!inHunk) {
      if (text.startsWith("Binary files")) binary = true;
      continue;
    }

    // "\ No newline at end of file" is a note about the line above, not a line.
    if (text.startsWith("\\")) {
      lines.push({ kind: "meta", text, oldLine: null, newLine: null });
      continue;
    }

    const marker = text[0];
    const body = text.slice(1);
    if (marker === "+") {
      stat.added++;
      lines.push({ kind: "add", text: body, oldLine: null, newLine: newLine++ });
    } else if (marker === "-") {
      stat.removed++;
      lines.push({ kind: "remove", text: body, oldLine: oldLine++, newLine: null });
    } else if (marker === " ") {
      lines.push({ kind: "context", text: body, oldLine: oldLine++, newLine: newLine++ });
    }
    // Anything else inside a hunk is a trailing blank from the final split.
  }

  return { lines, stat, empty: lines.length === 0, binary };
}

File: src-tauri/src/gitsync/diff.rs
//! Diffs, returned as git's own unified text.
//!
//! Unparsed on purpose: hunk splitting is what the viewer needs and nothing else
//! does, so it happens once in `utils/diff.ts` beside the component. Sending
//! structured hunks over IPC would mean a second representation of the same
//! thing, kept in step by hand.

use std::path::Path;

use super::{git, require_repo, stdout};
use crate::gitsync::status::{FileEntry, FileState};
use crate::AppResult;

/// `--no-color` and `--no-ext-diff` because the user's own `.gitconfig` may set a
/// pager, a colour scheme or an external difftool, none of which survive being
/// pasted into a `<pre>`. `-U3` pins the context so the viewer's line numbers do
/// not shift with whoever's config it is.
const DIFF_FLAGS: &[&str] = &["--no-color", "--no-ext-diff", "-U3"];

/// A working-tree file's diff — staged (`HEAD` vs index) or not (index vs disk).
pub fn diff_worktree_file(dir: &Path, path: &str, staged: bool) -> AppResult<String> {
    require_repo(dir)?;

    let mut args = vec!["diff"];
    if staged {
        args.push("--cached");
    }
    args.extend_from_slice(DIFF_FLAGS);
    args.push("--");
    args.push(path);

    let out = git(dir, &args)?;
    let text = String::from_utf8_lossy(&out.stdout).to_string();

    // An untracked file has nothing to diff against, so `git diff` says nothing.
    // Showing it as an addition is more use than showing an empty pane.
    if text.is_empty() && !staged {
        return untracked_as_addition(dir, path);
    }
    Ok(text)
}

/// `--no-index` compares two paths git does not track, so the whole file reads as
/// added. It exits 1 when they differ, which here is the expected outcome and not
/// a failure.
fn untracked_as_addition(dir: &Path, path: &str) -> AppResult<String> {
    let null = if cfg!(windows) { "NUL" } else { "/dev/null" };
    let mut args = vec!["diff", "--no-index"];
    args.extend_from_slice(DIFF_FLAGS);
    args.push("--");
    args.push(null);
    args.push(path);

    Ok(String::from_utf8_lossy(&git(dir, &args)?.stdout).to_string())
}

/// The files a commit touched, with the same state vocabulary the panel uses.
pub fn commit_files(dir: &Path, hash: &str) -> AppResult<Vec<FileEntry>> {
    require_repo(dir)?;

    // `--name-status` with `-m --first-parent` so a merge reports what it brought
    // in rather than nothing at all, which is what a bare merge diff produces.
    let out = git(
        dir,
        &[
            "show",
            "--name-status",
            "--format=",
            "-m",
            "--first-parent",
            "-z",
            hash,
        ],
    )?;
    if !out.status.success() {
        return Ok(Vec::new());
    }

    let raw = String::from_utf8_lossy(&out.stdout);
    let mut fields = raw.split('\0').filter(|f| !f.is_empty());
    let mut files = Vec::new();

    // `-z` emits status and path as separate fields, and a rename emits three:
    // status, old path, new path.
    while let Some(code) = fields.next() {
        let letter = code.chars().next().unwrap_or(' ');
        let Some(first) = fields.next() else { break };
        let path = if letter == 'R' || letter == 'C' {
            fields.next().unwrap_or(first)
        } else {
            first
        };
        files.push(FileEntry {
            path: path.to_string(),
            state: match letter {
                'A' => FileState::Added,
                'D' => FileState::Deleted,
                'R' | 'C' => FileState::Renamed,
                _ => FileState::Modified,
            },
        });
    }

    Ok(files)
}

/// One file's diff inside a commit.
pub fn diff_commit_file(dir: &Path, hash: &str, path: &str) -> AppResult<String> {
    require_repo(dir)?;

    let mut args = vec!["show", "--format=", "-m", "--first-parent"];
    args.extend_from_slice(DIFF_FLAGS);
    args.push(hash);
    args.push("--");
    args.push(path);

    Ok(stdout(&git(dir, &args)?))
}
